README.md
# NAME

Five::X - A set of modules that enable strict and warnings and features

# SYNOPSIS

```perl
#! /usr/bin/perl
use Five::Ten; # use feature ':5.10'; use strict; use warnings;
say "Turned on features for 5.10";
```

# DESCRIPTION

Either of these modules turns on strict, warnings and features in the calling package.

I have tried to document all the new features/experimentals per version. To
understand what a feature or experimental is, one should read the _perldelta_
document for that version.

## Experimental

Starting with perl version 5.18, some features are experimental and will warn in
the catagory `experimental::<feature>`. These will not be switched on by
default, you'll have to add the ':experimental' tag to enable these features and
disable the warnings that come with them.

## experimental\_versions

Returns an ordered list of versions that have warings in the `experimental::`
class without a feature category.

## experimental\_warnings($version)

Returns a listref of the warnings in the `experimental::` class for the given
perl version.

## features($version)

This returns a ref to a list of feature-names that are available in `$version`
that are not marked as experimental.

Keep in mind features can be removed (like **autoderef**) and will not show if
the actual version of perl doesn't have them anymore.

## all\_features

This returns a ref to a list of feature-names that ar available in `$version`,
including the features marked as experimental.

## unfeatures($version)

Newer versions of perl have features that one would like to turn off as the
syntax they represent is discouraged. All the `Five::X` do this by default. If
one wants indirect object syntax under [Five::ThirtyTwo](https://metacpan.org/pod/Five%3A%3AThirtyTwo) or higher one must
explicitly turn it on:

```perl
use feature 'indirect';
```

# SEE ALSO

[Five::Ten](https://metacpan.org/pod/Five%3A%3ATen) [Five::Twelve](https://metacpan.org/pod/Five%3A%3ATwelve) [Five::Fourteen](https://metacpan.org/pod/Five%3A%3AFourteen) [Five::Sixteen](https://metacpan.org/pod/Five%3A%3ASixteen) [Five::Eigtteen](https://metacpan.org/pod/Five%3A%3AEigtteen)

[Five::Twenty](https://metacpan.org/pod/Five%3A%3ATwenty) [Five::Twentytwo](https://metacpan.org/pod/Five%3A%3ATwentytwo) [Five::Twentyfour](https://metacpan.org/pod/Five%3A%3ATwentyfour) [Five::Twentysix](https://metacpan.org/pod/Five%3A%3ATwentysix) [Five::Twentyeight](https://metacpan.org/pod/Five%3A%3ATwentyeight)

[Five::Thirty](https://metacpan.org/pod/Five%3A%3AThirty) [Five::Thirtytwo](https://metacpan.org/pod/Five%3A%3AThirtytwo) [Five::Thirtyfour](https://metacpan.org/pod/Five%3A%3AThirtyfour) [Five::Thirtysix](https://metacpan.org/pod/Five%3A%3AThirtysix) [Five::Thirtyeight](https://metacpan.org/pod/Five%3A%3AThirtyeight)

# COPYRIGHT

© MMXXIII - Abe Timmerman <abeltje@cpan.org>

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Changes on 2015-06-17 for repository at:
  Fetch URL: ssh://git@source.test-smoke.org:9999/~/fivex
  Push  URL: ssh://git@source.test-smoke.org:9999/~/fivex

Enjoy!

________________________________________________________________________________
[9355a53] by abeltje on Wed, 17 Jun 2015 21:34:46 +0200

     More documentation.


M	lib/Five/Eight.pm
M	lib/Five/Eighteen.pm
M	lib/Five/Fourteen.pm
M	lib/Five/Sixteen.pm
M	lib/Five/Ten.pm
M	lib/Five/Twelve.pm
M	lib/Five/Twenty.pm
M	lib/Five/TwentyTwo.pm
M	lib/Five/X.pm
________________________________________________________________________________
[237ea26] by abeltje on Wed, 17 Jun 2015 08:03:52 +0200

     Autocommit for distribution Five::X 0.01_01 (test)


M	Changes
M	lib/Five/X.pm
________________________________________________________________________________
[06da06d] by abeltje on Wed, 17 Jun 2015 07:53:12 +0200

     Add a Changes file.


A	Changes
M	MANIFEST
________________________________________________________________________________
[4babda7] by abeltje on Wed, 17 Jun 2015 07:50:07 +0200

     Fix Makefile.PL for release-script


M	Makefile.PL
________________________________________________________________________________
[4070753] by abeltje on Wed, 17 Jun 2015 07:31:46 +0200

     TestFix: Test::Pod::Coverage and perlversion

        Because Test::Pod::Coverage needs to compile the module, we cannot 
        determine Pod-Coverage for modules that require a higher perl
        version than we are using during running of the test. Skip these
        modules.


M	xt/30-pod-coverage.t
________________________________________________________________________________
[9eb5615] by abeltje on Wed, 17 Jun 2015 07:05:25 +0200

     Add 5.22 + experimental:: warnings without feature

        Add support for 5.22 Five::TwentyTwo (with experimentals)

        Also extend the ':experimental' tag to versions before 5.18, when
        'experimental::' warnings without a feature were introduced. This
        means that «use Five::Ten ':experimental'» allows one to use
        'given/when' on 5.18 without warnings.


M	MANIFEST
M	lib/Five/Eighteen.pm
M	lib/Five/Fourteen.pm
M	lib/Five/Sixteen.pm
M	lib/Five/Ten.pm
M	lib/Five/Twelve.pm
M	lib/Five/Twenty.pm
A	lib/Five/TwentyTwo.pm
M	lib/Five/X.pm
A	t/5_022.t
M	xt/00-compile.t
________________________________________________________________________________
[d1df4ef] by abeltje on Sun, 7 Jun 2015 21:56:01 +0200

     Catch up until perl 5.20

        Still needs some tests for 5.16+


A	.gitignore
M	MANIFEST
A	MANIFEST.SKIP
M	Makefile.PL
M	lib/Five/Eight.pm
M	lib/Five/Eighteen.pm
M	lib/Five/Fourteen.pm
M	lib/Five/Sixteen.pm
M	lib/Five/Ten.pm
M	lib/Five/Twelve.pm
A	lib/Five/Twenty.pm
M	lib/Five/X.pm
M	t/5_008.t
M	t/5_010.t
M	t/5_012.t
M	t/5_014.t
M	t/5_016.t
M	t/5_018.t
A	t/5_020.t
A	xt/00-compile.t
A	xt/10-manifest.t
A	xt/20-pod.t
A	xt/30-pod-coverage.t
________________________________________________________________________________
[0dd9f65] by abeltje on Mon, 1 Jun 2015 06:22:33 +0200

     Update to 5.18

        Added support up to perl 5.018. Removed subversions, as they won't
        change the featuresets.


M	MANIFEST
M	Makefile.PL
A	lib/Five/Eighteen.pm
A	lib/Five/Fourteen.pm
A	lib/Five/Sixteen.pm
D	lib/Five/Ten/One.pm
D	lib/Five/Ten/Zero.pm
D	lib/Five/Twelve/One.pm
D	lib/Five/Twelve/Zero.pm
A	lib/Five/X.pm
D	t/5_010_000.t
D	t/5_010_001.t
D	t/5_012_000.t
D	t/5_012_001.t
A	t/5_014.t
A	t/5_016.t
A	t/5_018.t
________________________________________________________________________________
[1b9ae18] by abeltje on Sun, 26 Sep 2010 10:11:31 +0000

    
        * [CHANGE]
         + Add the distribution stuff.

         git-svn-id: http://fidovcs/repos/cpan/trunk/fivex@93
        03e79c58-f747-0410-a819-f71e7ca513d9


A	MANIFEST
________________________________________________________________________________
[7ed176b] by abeltje on Sun, 26 Sep 2010 10:11:15 +0000

    
        * [CHANGE]
         + Add the distribution stuff.

         git-svn-id: http://fidovcs/repos/cpan/trunk/fivex@92
        03e79c58-f747-0410-a819-f71e7ca513d9


A	Makefile.PL
________________________________________________________________________________
[7aea2bb] by abeltje on Sun, 26 Sep 2010 10:05:19 +0000

    
        * [CHANGE]
         + Initial import of files.

         git-svn-id: http://fidovcs/repos/cpan/trunk/fivex@91
        03e79c58-f747-0410-a819-f71e7ca513d9


A	lib/Five/Eight.pm
A	lib/Five/Ten.pm
A	lib/Five/Ten/One.pm
A	lib/Five/Ten/Zero.pm
A	lib/Five/Twelve.pm
A	lib/Five/Twelve/One.pm
A	lib/Five/Twelve/Zero.pm
A	t/5_008.t
A	t/5_010.t
A	t/5_010_000.t
A	t/5_010_001.t
A	t/5_012.t
A	t/5_012_000.t
A	t/5_012_001.t
________________________________________________________________________________
[381f0fb] by abeltje on Sun, 26 Sep 2010 07:54:03 +0000

    
        * [CHANGE]
         + like Modern::Perl, nicer name imo.

         git-svn-id: http://fidovcs/repos/cpan/trunk/fivex@90
        03e79c58-f747-0410-a819-f71e7ca513d9


lib/Five/X.pm
package Five::X;

our $VERSION = '0.02';

use Exporter 'import';
our @EXPORT = qw/ experimental_versions experimental_warnings /;

# These are experimental warnings without a feature, and backward incompatible!
# I'd like to say: use Five::Ten ':experimental';
# and be able to use given/when on 5.18
my %experimental = (
    '5.018' => [qw/experimental::smartmatch/],
    '5.020' => [qw/experimental::autoderef/],
    '5.022' => [qw/experimental::const_attr/],
);

sub experimental_versions {
    return sort keys %experimental;
}

sub experimental_warnings {
    my ($version) = @_;
    return [ ] if !exists $experimental{$version};
    return $experimental{$version};
}

=head1 NAME

FiveX - A set of modules that enable strict and warnings and features

=head1 SYNOPSIS

    #! /usr/bin/perl
    use Five::Ten; # use 5.010; use feature ':5.10'; use strict; use warnings;
    say "Turned on features for 5.10";

=head1 DESCRIPTION

Either of these modules turns on strict, warnings and features in the calling package.

=head2 use feature

Starting with perl version 5.18, some features are experimental. These will
not be switched on by default, you'll have to add the ':experimental' tag.

=head2 experimental_versions

Returns an ordered list of versions that have warings in the C<experimental::>
class without a feature category.

=head2 experimental_warnings($version)

Returns a listref of the warnings in the C<experimental::> class for the given
perl version.

=head3 5.018

=over

=item experimental::smartmatch  (feature :5.10)

=back

=head3 5.020

=over

=item experimental::smartmatch  (feature :5.10)

=item experimental::autoderef   (5.14)

=back

=head3 5.022

=over

=item experimental::smartmatch  (feature :5.10)

=item experimental::autoderef   (5.14)

=item experimental::const_attr  (5.22)

=back

=head1 COPYRIGHT

(c) MMXV - Abe Timmerman <abeltje@cpan.org>

=cut
